torch/csrc/api/include/torch/data/datasets/chunk.h

```diff
@@ -0,0 +1,455 @@
+  /// thread.
+  BatchType get_batch() {
+    std::unique_lock<std::mutex> lock(queue_mutex_);
+    cv_read_.wait(lock, [this] {
+      // wait till there is available data in the queue or if all chunks are
+      // loaded (i.e. the dataset is exhausted for this epoch)
+      return (
+          this->total_example_count_in_queue_ >= batch_size_ ||
+          this->remaining_chunk_count_ == 0);
+    });
+    if (batch_queue_.empty()) {
+      lock.unlock();
+      AT_ASSERT(remaining_chunk_count_ == 0);
+
+      // All batches have been retrieved. Return an empty batch.
+      return nullopt;
+    }
+
+    auto batch = batch_queue_.front();
+    batch_queue_.pop();
+    if (batch.exception) {
+      throw WorkerException(batch.exception);
+    }
+
+    total_example_count_in_queue_ -= batch.batch_data.size();
+    lock.unlock();
+    cv_write_.notify_all();
+
+    return batch.batch_data;
+  }
+
+  // skip one chunk
+  void skip_chunk() {
+    std::unique_lock<std::mutex> lock(queue_mutex_);
+    AT_ASSERT(remaining_chunk_count_ > 0);
+    remaining_chunk_count_--;
+    lock.unlock();
+    cv_read_.notify_all();
+  }
+
+  /// Push preloaded chunks to batch queue. Called from the ChunkDataset worker
+  /// threads.
+  void add_chunk_data(UnwrappedBatchType data) {
+    std::unique_lock<std::mutex> lock(queue_mutex_);
+    cv_write_.wait(lock, [this] {
+      // stop loading if we have preloaded enough data.
+      return this->total_example_count_in_queue_ < this->queue_capacity_ || stop_;
+    });
+
+    if (stop_){
+      // When stop_ is true, it means this current thread needs to be tore down.
+      // Return without any further processing.
+      lock.unlock();
+      return;
+    }
+
+    auto data_size = data.size();
+    auto remaining_size = data_size;
+    example_sampler_.reset(data_size);
+
+    auto fill_batch = [&](size_t example_count, UnwrappedBatchType& batch) {
+      auto batch_example_indices = this->example_sampler_.next(example_count);
+      AT_ASSERT(
+          batch_example_indices &&
+          batch_example_indices.value().size() == example_count)
+      BatchRequestType indices = batch_example_indices.value();
+      for (size_t i : indices) {
+        batch.emplace_back(std::move(data[i]));
+      }
+      remaining_size -= example_count;
+    };
+
+    if (!batch_queue_.empty()) {
+      // if the queue has existing data, and the last batch doesn't have enough
+      // examples to fill a batch_size batch, add more example to this batch first.
+      auto& batch = batch_queue_.back();
+      size_t current_count = batch.batch_data.size();
+      if (current_count < batch_size_) {
+        auto example_count =
+            std::min(remaining_size, batch_size_ - current_count);
+        fill_batch(example_count, batch.batch_data);
+      }
+    }
+
+    // If we still have data remaining after filling the last pushed batch, add
+    // them to the queue too.
+    while (remaining_size > 0) {
```

Comment:
I think you could compact the loop above with this one by doing something like.
```cpp
UnwrappedBatchType new_batch;
UnwrappedBatchType& current_batch = hasIncompleteBatch() ? batch_queue_.back() : new_batch;

// rest of the code stays almost the same, except that you have t
// account for current_batch being non-empty
```

Reply:
>rest of the code stays almost the same, except that you have to account for current_batch being non-empty

It actually has more checks: 
1. depending on hasIncompleteBatch(), whether to resize current_batch. We want to allocate a batch size upfront to avoid unnessary re-allocation during emplacing.
2. depending on hasIncompleteBatch(), how to caucl

The unify hurt the readability and unnecessarily increase the complexibility.

Reply:
This doesn't actually make the code shorter as it introduces more checks inside the loop. 

1. Depending on the return value of hasIncompleteBatch(), check if we need to reserve space for current_batch. We want to allocate a batch size upfront to avoid unnessary re-allocation during emplacing.

2. Depending on the return value of hasIncompleteBatch(), check if we need to push current_batch to the queue or leave it there. 

As it turns out, the unified code is actually longer.

Besides, the unified code has a slightly worse perf as it needs to access the queue to check hasIncompleteBatch() every time, while in reality, one time is enough.

So let's keep it as it is for now.

Reply:
Ok, fine.